core/src/graph/page.rs

```diff
@@ -407,4 +428,247 @@ mod test {
+		let encrypted_key =
+			ResolvedKeyPair { key_id: 1, key_pair: KeyPairType::Version1_0(StackKeyPair::gen()) };
+		let other_key =
+			ResolvedKeyPair { key_id: 2, key_pair: KeyPairType::Version1_0(StackKeyPair::gen()) };
+		let page_data = PageDataBuilder::new(Friendship(privacy_type))
+			.with_page(page_id, &connections, &prids, content_hash)
+			.with_encryption_key(encrypted_key)
+			.build();
+
+		// act
+		let graph_page = GraphPage::try_from((page_data.get(0).unwrap(), &dsnp, &vec![other_key]));
+
+		// assert
+		assert!(graph_page.is_err());
+	}
+
+	#[test]
+	fn removed_page_data_provider_should_return_removed_page_as_expected() {
+		// arrange
+		let graph = GraphPage {
+			page_id: 1,
+			privacy_type: PrivacyType::Private,
+			content_hash: 10,
+			prids: vec![DsnpPrid::from(vec![1u8, 2, 3, 4, 5, 6, 7, 8])],
+			connections: vec![DsnpGraphEdge { user_id: 70, since: 2873 }],
+		};
+		let expected = PageData { page_id: 1, content: vec![], content_hash: 10 };
+
+		// act
+		let removed = graph.to_removed_page_data();
+
+		// assert
+		assert_eq!(expected, removed);
+	}
+
+	#[test]
+	fn public_page_data_provider_should_return_public_page_as_expected() {
+		// arrange
+		let page_id = 10;
+		let privacy_type = Public;
+		let content_hash = 20;
+		let connections = vec![1, 2, 3, 4];
+		let page_data = PageDataBuilder::new(Follow(privacy_type))
+			.with_page(page_id, &connections, &vec![], content_hash)
+			.build();
+		let graph = GraphPage {
+			page_id,
+			privacy_type,
+			content_hash,
+			prids: vec![],
+			connections: connections
+				.iter()
+				.map(|c| DsnpGraphEdge { user_id: *c, since: 0 })
+				.collect(),
+		};
+
+		// act
+		let public = graph.to_public_page_data();
+
+		// assert
+		assert!(public.is_ok());
+		let public = public.unwrap();
+		assert_eq!(&public, page_data.get(0).unwrap());
+	}
+
+	#[test]
+	fn private_page_data_provider_should_return_private_page_as_expected() {
+		// arrange
+		let page_id = 10;
+		let privacy_type = PrivacyType::Private;
+		let content_hash = 20;
+		let dsnp = DsnpVersionConfig::new(DsnpVersion::Version1_0);
+		let connections = vec![1, 2, 3, 4];
+		let prids: Vec<DsnpPrid> = connections.iter().map(|id| DsnpPrid::from(*id)).collect();
+		let key =
+			ResolvedKeyPair { key_id: 1, key_pair: KeyPairType::Version1_0(StackKeyPair::gen()) };
+		let page_data = PageDataBuilder::new(Friendship(privacy_type))
+			.with_page(page_id, &connections, &prids, content_hash)
+			.with_encryption_key(key.clone())
+			.build();
+		let graph_page =
+			GraphPage::try_from((page_data.get(0).unwrap(), &dsnp, &vec![key.clone()]))
+				.expect("should work");
+
+		// act
+		let private = graph_page.to_private_page_data(&dsnp, &key).expect("should work");
+		let graph_page2 = GraphPage::try_from((&private, &dsnp, &vec![key.clone()]));
+
+		// assert
+		assert!(graph_page2.is_ok());
+		let graph_page2 = graph_page2.unwrap();
+		assert_eq!(graph_page, graph_page2);
+	}
```

Comment:
Nit: do we have some non happy path tests alread? 
Looks good catching up all the work that went here

Reply:
We have some but it's not comprehensive yet. Most of these tests in this PR to verify the happy path implementation but we have another ticket for more end to end and comprehensive testing.